### PR TITLE
Fix panic related to update with refresh

### LINKED ISSUE #5946: Panic with pulumi up when using refresh
Sorry for not a precise issue, but I don't have time.

I experienced this issue today with my pulumi script:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x98 pc=0xeaa730]

goroutine 99 [running]:
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepGenerator).AnalyzeResources.func1(0xc001728270, 0xc6, 0xc0013f6240, 0xc0006b1790)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_generator.go:1301 +0x130
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*resourceMap).mapRange.func1(0x1b305a0, 0xc000bced70, 0x190a5e0, 0xc0013f6240, 0x1)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/deployment.go:126 +0x6b
sync.(*Map).Range(0xc0000a0f00, 0xc000673810)
	/Users/runner/hostedtoolcache/go/1.15.6/x64/src/sync/map.go:345 +0x1d9
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*resourceMap).mapRange(0xc0000a0f00, 0xc000673970)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/deployment.go:125 +0x55
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepGenerator).AnalyzeResources(0xc0014a17a0, 0x2c1fd80, 0x0)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/step_generator.go:1289 +0xac
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*deploymentExecutor).Execute(0xc0005eaea0, 0x20075a0, 0xc00172ef90, 0x7f9838b98118, 0xc0007e5180, 0x7fffffff, 0x1, 0x2c1fd80, 0x0, 0x0, ...)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/deployment_executor.go:277 +0xae6
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*Deployment).Execute(...)
	/Users/runner/work/pulumi/pulumi/pkg/resource/deploy/deployment.go:383
github.com/pulumi/pulumi/pkg/v2/engine.(*deployment).run.func1(0x2011fe0, 0xc0007e5180, 0xc0005f83c0, 0x20075a0, 0xc00172ef90, 0xc00172ef01, 0xc00059aa60, 0xc00004f4a0)
	/Users/runner/work/pulumi/pulumi/pkg/engine/deployment.go:251 +0x267
created by github.com/pulumi/pulumi/pkg/v2/engine.(*deployment).run
	/Users/runner/work/pulumi/pulumi/pkg/engine/deployment.go:237 +0x306
Finished (TagStatus.Failed) 'DeployInfra' in 00:00:12.3009585


```

It only happens when I do `up` with `-r`. Removing `-r` and deploy works again. Not sure what is going on.

I'm using the 2.15.6 version of pulumi.

### PR DESCRIPTION
**Pull request by lblackstone**
Fix #5946

### CODE REVIEW

File: pkg/engine/lifeycletest/refresh_test.go [hunk 1]
```diff
@@ -813,3 +813,49 @@ func TestRefreshStepWillPersistUpdatedIDs(t *testing.T) {
 		}
 	}
 }
+
+// TestRefreshUpdateWithDeletedResource validates that the engine handles a deleted resource without error on an
+// update with refresh.
+func TestRefreshUpdateWithDeletedResource(t *testing.T) {
```
Comment (pgavlin) [inline pkg/engine/lifeycletest/refresh_test.go:819]: We might need to add an analyzer to this test in order to properly validate the originally failing codepath. I don't think we've done that before, but it should be doable. LMK if you'd like some pointers on where to start.

Comment (lblackstone) [reply pkg/engine/lifeycletest/refresh_test.go:819]: This test panics without the fix applied, but not 100% sure if it's the same spot.

Comment (lblackstone) [reply pkg/engine/lifeycletest/refresh_test.go:819]: Ok, confirmed that the test panics in the same spot, so this should be good to go.

```
=== RUN   TestRefreshUpdateWithDeletedResource
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x98 pc=0x19f6710]

goroutine 55 [running]:
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepGenerator).AnalyzeResources.func1(0xc000270f00, 0x35, 0xc0004e4480, 0xc0003d7600)
	/Users/levi/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/step_generator.go:1301 +0x130
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*resourceMap).mapRange.func1(0x25347a0, 0xc0002491e0, 0x2405ac0, 0xc0004e4480, 0x1)
	/Users/levi/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/deployment.go:127 +0x6b
sync.(*Map).Range(0xc000591200, 0xc0006a7810)
	/usr/local/Cellar/go/1.15.5/libexec/src/sync/map.go:345 +0x1d9
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*resourceMap).mapRange(0xc000591200, 0xc0006a7970)
	/Users/levi/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/deployment.go:126 +0x55
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*stepGenerator).AnalyzeResources(0xc0004e4a20, 0x0, 0x0)
	/Users/levi/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/step_generator.go:1289 +0xac
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*deploymentExecutor).Execute(0xc00000e820, 0x2938e60, 0xc000591290, 0x9c90e58, 0xc0000e2c80, 0x0, 0x1, 0x0, 0x0, 0x0, ...)
	/Users/levi/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/deployment_executor.go:277 +0xae6
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*Deployment).Execute(...)
	/Users/levi/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/deployment.go:384
github.com/pulumi/pulumi/pkg/v2/engine.(*deployment).run.func1(0x293dd20, 0xc0000e2c80, 0xc0000e2000, 0x2938e60, 0xc000591290, 0xc000591201, 0xc000248ec0, 0xc000389320)
	/Users/levi/go/src/github.com/pulumi/pulumi/pkg/engine/deployment.go:252 +0x287
created by github.com/pulumi/pulumi/pkg/v2/engine.(*deployment).run
	/Users/levi/go/src/github.com/pulumi/pulumi/pkg/engine/deployment.go:237 +0x306
```

**Comment by lblackstone** [comment]
@pgavlin If this looks good, feel free to merge.

**Pull request merged**